File: 14-queue-using-stacks/solution.cpp
// Implement Queue using Two Stacks
// Time: O(1) amortized for push/pop, Space: O(n)

#include <iostream>
#include <stack>
using namespace std;

class MyQueue {
private:
    stack<int> stackIn;
    stack<int> stackOut;
    
    void move() {
        if (stackOut.empty()) {
            while (!stackIn.empty()) {
                stackOut.push(stackIn.top());
                stackIn.pop();
            }
        }
    }
    
public:
    MyQueue() {}
    
    void push(int x) {
        stackIn.push(x);
    }
    
    int pop() {
        move();
        int val = stackOut.top();
        stackOut.pop();
        return val;
    }
    
    int peek() {
        move();
        return stackOut.top();
    }
    
    bool empty() {
        return stackIn.empty() && stackOut.empty();
    }
};

int main() {
    MyQueue q;
    q.push(1);
    q.push(2);
    cout << q.peek() << endl;  // 1
    cout << q.pop() << endl;   // 1
    cout << boolalpha << q.empty() << endl; // false
    return 0;
}

File: 14-queue-using-stacks/solution.java
// Implement Queue using Two Stacks
// Time: O(1) amortized for push/pop, Space: O(n)

import java.util.*;

class MyQueue {
    private Stack<Integer> stackIn;
    private Stack<Integer> stackOut;
    
    public MyQueue() {
        stackIn = new Stack<>();
        stackOut = new Stack<>();
    }
    
    public void push(int x) {
        stackIn.push(x);
    }
    
    public int pop() {
        move();
        return stackOut.pop();
    }
    
    public int peek() {
        move();
        return stackOut.peek();
    }
    
    public boolean empty() {
        return stackIn.isEmpty() && stackOut.isEmpty();
    }
    
    private void move() {
        if (stackOut.isEmpty()) {
            while (!stackIn.isEmpty()) {
                stackOut.push(stackIn.pop());
            }
        }
    }
    
    public static void main(String[] args) {
        MyQueue q = new MyQueue();
        q.push(1);
        q.push(2);
        System.out.println(q.peek());  // 1
        System.out.println(q.pop());   // 1
        System.out.println(q.empty()); // false
    }
}
